p2p/resource_manager.go

```diff
@@ -0,0 +1,161 @@
+package p2p
+
+import (
+	"math"
+
+	"github.com/harmony-one/harmony/internal/utils"
+
+	humanize "github.com/dustin/go-humanize"
+	libp2p "github.com/libp2p/go-libp2p"
+	network "github.com/libp2p/go-libp2p/core/network"
+	rcmgr "github.com/libp2p/go-libp2p/p2p/host/resource-manager"
+	memory "github.com/pbnjay/memory"
+)
+
+// GetNumFDs returns the File Descriptors as MaxInt.
+func GetNumFDs() uint64 {
+	return uint64(math.MaxUint64)
+}
+
+// mostly copy/pasted from https://github.com/ipfs/rainbow/blob/main/rcmgr.go
+// which is itself copy-pasted from Kubo, because libp2p does not have
+// a sane way of doing this.
+func makeResourceMgr(enabled bool, maxMemory, maxFD uint64, connMgrHighWater int) (network.ResourceManager, error) {
+	if !enabled {
+		rmgr, err := rcmgr.NewResourceManager(rcmgr.NewFixedLimiter(rcmgr.InfiniteLimits))
+		utils.Logger().Info().Msg("go-libp2p Resource Manager DISABLED")
+		return rmgr, err
+	}
+
+	// Auto-scaled limits based on available memory/fds.
+	if maxMemory == 0 {
+		maxMemory = uint64((float64(memory.TotalMemory()) * 0.25))
+		if maxMemory < 1<<20 { // 1 GiB
+			maxMemory = 1 << 20
+		}
```

Comment:
So here we are reserving 25% of whole RAM, but if the RAM < 1GB - we will put here 1 GB as a limit, this one looks a bit tricky imho.

From one side:
* we will need to remember that we must use machines with > 1GB of RAM or linux OOM killer will find us...
From other side:
* I believe this is okay, rn we have 2 or 4 GB machines

cc: @sophoah lets just have this in your mind for now

Reply:
Agreed, @GheisMohammadi let's put 0.5GB as minimal amount of RAM

Reply:
by the way, i think that PR is good already.  1 << 20 = 1 MiB, not 1 GiB.

Reply:
yes, you are right, thanks for fixing comment @sophoah, let's merge it